io/io/src/TFileMergeServer.cxx

```diff
@@ -0,0 +1,348 @@
+   {
+      // Merge the current inputs into the output file.
+
+      // Remove object that can *not* be incrementally merged
+      // and will *not* be reset by the client code.
+      R__DeleteObject(fMerger.GetOutputFile(), false);
+
+      for (unsigned int f = 0; f < fClients.size(); ++f) {
+         fMerger.AddFile(fClients[f].fFile);
+      }
+
+      Bool_t result = fMerger.PartialMerge(TFileMerger::kAllIncremental);
+
+      // Remove any 'resetable' object (like TTree) from the input file
+      // so that they will not be re-merged. Keep only the object that
+      // always need to be re-merged (Histograms).
+
+      for (unsigned int f = 0; f < fClients.size(); ++f) {
+         if (fClients[f].fFile) {
+            R__DeleteObject(fClients[f].fFile, true);
+         } else {
+            // We back up the file (probably due to memory constraint)
+            TFile *file = TFile::Open(fClients[f].fLocalName, "UPDATE");
+            // Remove object that can be incrementally merged and
+            // will be reset by the client code.
+            R__DeleteObject(file, true);
+            file->Write();
+            delete file;
+         }
+      }
+      fLastMerge = TTimeStamp();
+      fNClientsContact = 0;
+      fClientsContact.Clear();
+
+      return result;
+   }
+
+   void RegisterClient(UInt_t clientId, TFile *file)
+   {
+      ++fNClientsContact;
+
+      fClientsContact.SetBitNumber(clientId);
+
+      if (fClients.size() < clientId + 1) {
+         fClients.push_back(ClientInfo(fFilename, clientId));
+      }
+      fClients[clientId].Set(file);
+   }
+};
+
+} // unnamed namespace
+
+namespace ROOT {
+namespace Experimental {
+
+std::shared_ptr<TFileMergeClient> TFileMergeServer::GetClient()
+{
+   std::shared_ptr<TFileMergeClient> client(new TFileMergeClient(*this));
+   fClients.push_back(client);
+   return client;
+}
+
+void TFileMergeServer::Listen()
+{
+   std::unique_lock<std::mutex> wlock(fWriteMutex);
+
+   THashTable mergers;
+   bool cache = false;
+
+   static Int_t clientId = 0;
+
+   while (!finish) {
+      fCV.wait(wlock, [this](){ return !this->fQueue.empty(); });
+
+      while (!fQueue.empty()) {
+         TBufferFile *buffer;
+
+         {
+            std::lock_guard<std::mutex> qlock(fQueueMutex);
+            buffer = fQueue.front();
+            fQueue.pop();
+         }
+
+         buffer->SetReadMode();
+         buffer->SetBufferOffset();
+
+         Long64_t length;
+         TString filename;
+
+         buffer->ReadTString(filename);
+         buffer->ReadLong64(length);
+
+         // UPDATE because we need to remove the TTree after merging them.
+         TMemFile *transient = new TMemFile(filename, buffer->Buffer() + buffer->Length(), length, "UPDATE");
+
+         buffer->SetBufferOffset(buffer->Length() + length);
+
+         // control how often the histogram are merged.  Here as soon as half the clients have reported.
+         const Float_t clientThreshold = 0.75;
```

Comment:
yes, it should be configurable by the user.  This is copied from  the tutorial tutorials/net/parallelMergeServer.C where this was an expedient (the whole server is a single script with no user interface per se, just tweak the source and run).  So in a future update we ought to expose this  .... however this mechanism is really sub-optimal for histogram.